### PR TITLE
Add feature to revoke bindings

### LINKED ISSUE #3971: [Enhancement] Revoke key or mouse button binding
This is an issue/idea I originally thought related to button mapping and thus was inserted as a comment in another issue. My issue is exclusively with mouse bindings, and it seemed possible using "seat" but I have not had any luck. Nonetheless, it applies to keyboard bindings as well. Is it possible to have Sway revoke or abandon a bindsym?

Use case: if I don't bind button8 or button9 at all in the config, applications like nemo correctly process them as "back" and "forward" buttons. If I do create a binding in the config (or with swaymsg 'bindsym <blah blah>' while running) like:
`bindsym --whole-window button8 exec virsh send-key Arch KEY_LEFTALT KEY_LEFT`

`bindsym --whole-window button9 exec virsh send-key Arch KEY_LEFTALT KEY_RIGHT`
It works exactly as it should- pressing button8 and button9 executes the respective virsh command and virsh sends the keys to my VM. Of course now nemo isn't going to see button8 or button9 since Sway is executing the virsh commands. I can't seem to get button8 and button9 to unmap- I can bindsym them to a different command, but they will not be recognized by nemo again unless I exit sway and start again (again with no bindsym commands for button8 or button9 in the config). xev doesn't even register them being pushed, though libinput events does.

I tried playing with seat - cursor like:
`swaymsg 'bindsym --input-device=blahblah --whole-window button8 seat - cursor press button8`

`...etc`
but that doesn't seem to set it back to how it would be if I hadn't bindsym'd button8 or button9 at all.

Im not experiencing a crash and the corresponding entries in the config (in the case where they are bound upon Sway's start) are listed above, so I don't think I need a Sway log or config post. I am running Sway version 1.0 from the Arch Community repo, but I can move to the latest git of wlroots or sway for testing, troubleshooting, etc. If a log or config is required, I'll be happy to upload them...

### PR DESCRIPTION
**Pull request by ForTheReallys**
This PR implements the feature request #3971 that will allow a user to revoke key, mouse, or switch bindings both for sway and swaybar.

Status:
- [x] Add unbindsym/unbindcode commands for sway
- [x] Add unbindsym/unbindcode commands for swaybar
- [x] Add unbindswitch command
- [x] Cleanup

Closes #3971

### CODE REVIEW

**Comment by RedSoxFan** [comment]
Before you get too deep, I think it would be better to just add a `--remove` or `--unbind` flag instead of a separate command. Adding the flag will heavily reduce code duplication and maintenance since every time the `bind` version get updated, the `unbind` version would likely also need mirroring changes.

**Comment by ForTheReallys** [comment]
Originally to clean everything up, I was going to create another function to "initialize" the bindings, which would get rid of all that duplicated code. Would you still prefer `--unbind` or `--remove`?

**Comment by RedSoxFan** [comment]
As long as there isn't duplication, I'm fine with either then

**Comment by ForTheReallys** [comment]
@RedSoxFan The latest commit implements some of the cleanup I was talking about to get rid of duplicated code.  It gets rid of all duplicated code except command line processing.  Let me know if you would still prefer another flag instead of another command.

**Comment by RedSoxFan** [comment]
If you can remove the flag duplication as well, the unbind family of commands is fine with me

**Comment by ForTheReallys** [comment]
Will do. Sorry for taking so long. This week has been way busier than I expected. This should be ready for review by tomorrow.

**Comment by ForTheReallys** [comment]
This is finally ready for review.

File: sway/commands/bind.c [hunk 1]
```diff
@@ -196,12 +196,122 @@ static struct cmd_results *identify_key(const char* name, bool first_key,
...
+			mode_bindings->items[i] = binding;
+			overwritten = true;
+		}
+	}
+
+	if (!overwritten) {
+		list_add(mode_bindings, binding);
+	}
+
+	sway_log(SWAY_DEBUG, "%s - Bound %s to command `%s`",
```
Comment (RedSoxFan) [inline sway/commands/bind.c]: I think this should be in the `!overwritten` body. The `Overwriting binding` log message should be sufficient.

File: sway/commands/bind.c [hunk 2]
```diff
@@ -196,12 +196,122 @@ static struct cmd_results *identify_key(const char* name, bool first_key,
...
+					bindtype, switchcombo);
+			return cmd_results_new(CMD_SUCCESS, NULL);
+		}
+	}
+
+	free_switch_binding(binding);
+	return cmd_results_new(CMD_FAILURE, "Could not find switch binding `%s`", switchcombo);
+}
+
+static struct cmd_results *binding_add_to_list(struct sway_binding *binding, list_t *mode_bindings,
```
Comment (RedSoxFan) [inline sway/commands/bind.c]: style: 80 column limit

File: sway/commands/bind.c [hunk 3]
```diff
@@ -196,12 +196,122 @@ static struct cmd_results *identify_key(const char* name, bool first_key,
...
+			mode_bindings->items[i] = binding;
+			overwritten = true;
+		}
+	}
+
+	if (!overwritten) {
+		list_add(mode_bindings, binding);
+	}
+
+	sway_log(SWAY_DEBUG, "%s - Bound %s to command `%s` for device '%s'",
```
Comment (RedSoxFan) [inline sway/commands/bind.c]: I think this should be in the `!overwritten` body. The `Overwriting binding` log message should be sufficient.

File: sway/commands/bind.c [hunk 4]
```diff
@@ -196,12 +196,122 @@ static struct cmd_results *identify_key(const char* name, bool first_key,
...
+			bindtype, keycombo, binding->command, binding->input);
+
+	return cmd_results_new(CMD_SUCCESS, NULL);
+}
+
+static struct cmd_results *binding_remove_from_list(struct sway_binding *binding, list_t *mode_bindings,
```
Comment (RedSoxFan) [inline sway/commands/bind.c]: style: 80 column limit

File: sway/commands/bind.c [hunk 5]
```diff
@@ -196,12 +196,122 @@ static struct cmd_results *identify_key(const char* name, bool first_key,
...
+			free_sway_binding(config_binding);
+			free_sway_binding(binding);
+			list_del(mode_bindings, i);
+			sway_log(SWAY_DEBUG, "%s - Unbound `%s` from device '%s'",
+					bindtype, keycombo, binding->input);
+			return cmd_results_new(CMD_SUCCESS, NULL);
+		}
+	}
+	free_sway_binding(binding);
+	return cmd_results_new(CMD_FAILURE, "Could not find binding `%s` for the given flags", keycombo);
```
Comment (RedSoxFan) [inline sway/commands/bind.c]: style: 80 column limit

File: sway/commands/bind.c [hunk 6]
```diff
@@ -196,12 +196,122 @@ static struct cmd_results *identify_key(const char* name, bool first_key,
...
+}
+
 static struct cmd_results *cmd_bindsym_or_bindcode(int argc, char **argv,
-		bool bindcode) {
-	const char *bindtype = bindcode ? "bindcode" : "bindsym";
+		bool bindcode, bool unbind) {
+	const char *bindtype;
+	int minargs = 2;
+
```
Comment (RedSoxFan) [inline sway/commands/bind.c]: Unnecessary blank line

File: sway/commands/bind.c [hunk 7]
```diff
@@ -248,15 +358,13 @@ static struct cmd_results *cmd_bindsym_or_bindcode(int argc, char **argv,
 			BINDING_MOUSECODE : BINDING_MOUSESYM;
 	}
 
-	if (argc < 2) {
+	if (argc < minargs) {
 		free_sway_binding(binding);
 		return cmd_results_new(CMD_FAILURE,
 			"Invalid %s command "
-			"(expected at least 2 non-option arguments, got %d)", bindtype, argc);
+			"(expected at least %d non-option arguments, got %d)", bindtype, minargs, argc);
```
Comment (RedSoxFan) [inline sway/commands/bind.c]: style: 80 column limit

File: sway/commands/bind.c [hunk 8]
```diff
@@ -310,47 +417,25 @@ static struct cmd_results *cmd_bindsym_or_bindcode(int argc, char **argv,
...
-					"'%s' for device '%s' to `%s` from `%s`",
-					argv[0], binding->input, binding->command,
-					config_binding->command);
-			}
-			free_sway_binding(config_binding);
-			mode_bindings->items[i] = binding;
-			overwritten = true;
-		}
+	if (unbind) {
+		return binding_remove_from_list(binding, mode_bindings, bindtype, argv[0]);
```
Comment (RedSoxFan) [inline sway/commands/bind.c]: style: 80 column limit

File: sway/commands/bind.c [hunk 9]
```diff
@@ -310,47 +417,25 @@ static struct cmd_results *cmd_bindsym_or_bindcode(int argc, char **argv,
...
-
-struct cmd_results *cmd_bindcode(int argc, char **argv) {
-	return cmd_bindsym_or_bindcode(argc, argv, true);
+	binding->command = join_args(argv + 1, argc - 1);
+	binding->order = binding_order++;
+	return binding_add_to_list(binding, mode_bindings, bindtype, argv[0], warn);
 }
 
-struct cmd_results *cmd_bindswitch(int argc, char **argv) {
+struct cmd_results *cmd_bind_or_unbind_switch(int argc, char **argv, bool unbind) {
```
Comment (RedSoxFan) [inline sway/commands/bind.c]: style: 80 column limit

**Comment by ForTheReallys** [comment]
I'm not entirely sure how I can make it much smaller.  In any case, I just forced push the branch with only three commits that addressed all your comments.  Let me know if I missed anything or if you need me to change anything else.  Thanks!

File: sway/sway.5.scd [hunk 10]
```diff
@@ -353,6 +353,14 @@ runtime.
 	*bindcode* [--whole-window] [--border] [--exclude-titlebar] [--release] [--locked] [--input-device=<device>] [--no-warn] <code> <command>
 	is also available for binding with key/button codes instead of key/button names.
 
+*unbindsym* [--whole-window] [--border] [--exclude-titlebar] [--release] [--input-device=<device>] <key combo>
```
Comment (RedSoxFan) [inline sway/sway.5.scd]: Can you move this down so it is in alphabetical order? _I know there are some entries that do not comply with this, but we try to keep them in order_

File: sway/sway.5.scd [hunk 11]
```diff
@@ -375,6 +383,9 @@ runtime.
 		bindswitch lid:toggle exec echo "Lid moved"
 ```
 
+*unbindswitch* <switch>:<state>
```
Comment (RedSoxFan) [inline sway/sway.5.scd]: Ditto here for alphabetical order

File: sway/commands/bar/bind.c [hunk 12]
```diff
@@ -9,10 +9,70 @@
...
+		list_t *mode_bindings) {
+	const char *name = get_mouse_button_name(binding->button);
+	for (int i = 0; i < mode_bindings->length; i++) {
+		struct bar_binding *other = mode_bindings->items[i];
+		if (other->button == binding->button &&
+				other->release == binding->release) {
+			free_bar_binding(other);
+			free_bar_binding(binding);
+			list_del(mode_bindings, i);
+			sway_log(SWAY_DEBUG, "[bar %s] Unbound binding for %u (%s)%s",
```
Comment (RedSoxFan) [inline sway/commands/bar/bind.c]: This debug log needs to be moved before `binding` is freed. `binding->button` and `binding->release` will be use-after-frees with the current version.

File: sway/commands/bind.c [hunk 13]
```diff
@@ -196,12 +196,123 @@ static struct cmd_results *identify_key(const char* name, bool first_key,
...
+static struct cmd_results *binding_remove(struct sway_binding *binding,
+		list_t *mode_bindings, const char *bindtype,
+		const char *keycombo) {
+	for (int i = 0; i < mode_bindings->length; ++i) {
+		struct sway_binding *config_binding = mode_bindings->items[i];
+		if (binding_key_compare(binding, config_binding)) {
+			free_sway_binding(config_binding);
+			free_sway_binding(binding);
+			list_del(mode_bindings, i);
+			sway_log(SWAY_DEBUG, "%s - Unbound `%s` from device '%s'",
```
Comment (RedSoxFan) [inline sway/commands/bind.c]: There is also a use-after-free here for `binding->input`

**Comment by ForTheReallys** [comment]
Thanks for catching that.  I can't believe I didn't see that.

**Comment by RedSoxFan** [comment]
Can you rebase the use-after-free fixes into their respective command commits for a cleaner commit history? Should be good to merge after that.

**Comment by RedSoxFan** [comment]
Thanks!

**Pull request merged**